### PR TITLE
Add migrator for `nom_analysis_activity` ID

### PR DESCRIPTION
**Pull request by JamesTessmer**
This migrator will be used to append version numbers to IDs that currently have none.

Add to resolve Issue https://github.com/microbiomedata/nmdc-schema/issues/2010

### CODE REVIEW

File: nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+        >>> m = Migrator()
+        >>> m.update_nom_analysis_activity_id({'id': '1'}) # test: adds version number to id
+        {'id': '1.1'}
+        """
+        self.logger.info(f"Processing NomAnalysisActivity: {nom_analysis_activity['id']}")
+
+        
+        # If this data object's `data_object_type` consists of the invalid value, replace it with the valid value.
+        if "." not in nom_analysis_activity['id']:
```
Comment (eecavanna) [inline nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py:30]: I am wondering whether this is a sufficient check. In other words, I am wondering whether it's possible that a "non-versioned" ID still contains a period for an unrelated reason.

File: nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        )
+
+    def update_nom_analysis_activity_id(self, nom_analysis_activity: dict) -> dict:
+        r"""
+        Updates a nom_analysis_activity's id to contain a version number (.1),
+        if it has no version number.
+
+        >>> m = Migrator()
+        >>> m.update_nom_analysis_activity_id({'id': '1'}) # test: adds version number to id
+        {'id': '1.1'}
```
Comment (eecavanna) [inline nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py:24]: Will you add an example for where the document's ID already ends with `.1`?

```suggestion
        {'id': '1.1'}
        >>> m.update_nom_analysis_activity_id({'id': '1.1'}) # test: no change
        {'id': '1.1'}
```

File: nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        """
+        self.logger.info(f"Processing NomAnalysisActivity: {nom_analysis_activity['id']}")
+
+        
+        # If this data object's `data_object_type` consists of the invalid value, replace it with the valid value.
+        if "." not in nom_analysis_activity['id']:
+            nom_analysis_activity['id'] = str(nom_analysis_activity['id']) + ".1"
+
+        return nom_analysis_activity
```
Comment (eecavanna) [inline nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py:33]: I think one of the requirements was to put the original `id` into a field named `alternative_identifiers`.

**Review by eecavanna — COMMENTED** [review]
Hi @JamesTessmer, thanks for implementing this. I left some feedback comments.

File: nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py (hunk 1 above)
Comment (eecavanna) [reply nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py:30]: Here's a regex pattern you can use to determine whether the ID "ends with a dot followed by one or more numeric digits":

```py
pattern = re.compile(r"\.[0-9]+$")  # ends with dot and number(s)
```

File: nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py (hunk 2 above)
Comment (eecavanna) [reply nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py:24]: I'd also like to see a counterexample where the document's ID contains a dot that is not at the end. That would serve as documentation about the migrator's behavior for such a corner case.

For example:

```py
>>> m.update_nom_analysis_activity_id({'id': 'nmdc:foo-12.3-potato'}) # test: no change
{'id': 'nmdc:foo-12.3-potato'}
```

File: nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py (hunk 3 above)
Comment (eecavanna) [reply nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py:33]: Here's how I did it in the example one: https://gist.github.com/eecavanna/fc1e20dc919f87746c8668e86ea68a4a#file-migrator_from_10_3_0_to_10_4_0-py-L42-L61

File: nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+        self.adapter.process_each_document(
+            "nom_analysis_activity_set", [self.update_nom_analysis_activity_id]
+        )
+
+    def update_nom_analysis_activity_id(self, nom_analysis_activity: dict) -> dict:
+        r"""
+        Updates a nom_analysis_activity's id to contain a version number (.1),
+        if it has no version number.
+
```
Comment (eecavanna) [inline nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py:21]: Great description!

File: nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        >>> m = Migrator()
+        >>> m.update_nom_analysis_activity_id({'id': '1'}) # test: adds version number to id
+        {'id': '1.1'}
+        """
+        self.logger.info(f"Processing NomAnalysisActivity: {nom_analysis_activity['id']}")
+
+        
+        # If this data object's `data_object_type` consists of the invalid value, replace it with the valid value.
```
Comment (eecavanna) [inline nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py:29]: I think this comment is out-of-date.

**Comment by JamesTessmer** [comment]
> Hi @JamesTessmer, thanks for implementing this. I left some feedback comments.

Thanks for all the feedback @eecavanna ! I'll get this all done today.

**Comment by JamesTessmer** [comment]
@eecavanna I've updated everything with the examples you showed and your suggestions.
I'm marking this as ready for review

File: nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            "nom_analysis_activity_set", [self.populate_alternative_identifiers,self.update_nom_analysis_activity_id]
+        )
+    
+    def is_valid_id(self, s: str) -> bool:
+        r"""
+        Helper function that checks whether a string is a valid ID (has a version number).
+        Reference: https://docs.python.org/3/library/re.html#re.Pattern.search
+
+        >>> m = Migrator()
+        >>> m.is_valid_id("invalid_id")
```
Comment (aclum) [inline nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py:28]: I think this will work but next time we should use the materialized patterns generated by running `make all` which generates `project/nmdc_materialized_patterns.yaml` (^(nmdc):wfnom-([0-9][a-z]{0,6}[0-9])-([A-Za-z0-9]{1,})(\.[0-9]{1,})$)

Comment (JamesTessmer) [reply nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py:28]: Ah, I didn't know about the patterns, thanks for pointing that out!
I can add that pattern to the regex instead.

Comment (JamesTessmer) [reply nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py:28]: Hmm, when I plug that in to the regex and test it says IDs without version numbers are still valid.

Comment (JamesTessmer) [reply nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py:28]: Ah, the issue is a period character is a wildcard so it's just seeing that. If I escape it with a backslash it does work as intended.

File: nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+from nmdc_schema.migrators.migrator_base import MigratorBase
+
+
+class Migrator(MigratorBase):
+    r"""Migrates a database between two schemas.
+    """
+
+    _from_version = "10.3.0"
+    _to_version = "10.4.0"
```
Comment (eecavanna) [inline nmdc_schema/migrators/migrator_from_10_3_0_to_10_4_0.py:11]: FYI @aclum: These lines, and the name of the migrator Python file, are the two places that would require changing if you wanted this migrator to indicate that it was associated with different schema versions.

**Comment by eecavanna** [comment]
I will update this branch to reference different schema version numbers. New commit(s) coming soon.

**Comment by eecavanna** [comment]
@JamesTessmer, do you know what the schema version is in which the **schema changes** associated with this migrator were introduced in? Alternatively, do you know what the PR is in which those schema changes were introduced into the repo? I'm trying to determine what the correct "to" and "from" versions are for this migrator.

CC: @turbomam

**Comment by eecavanna** [comment]
@JamesTessmer, do you know the version number of the schema release in which the schema changes associated with this migrator were introduced? Alternatively, do you know which PR those schema changes were introduced into the repo in? I'm trying to determine what the correct "to" and "from" versions are for this migrator. I assumed the 10.3.0 and 10.4.0 (which I have since replaced) were placeholders.

CC: @turbomam

**Comment by JamesTessmer** [comment]
I have no idea where the changes for these come from. In the issue (https://github.com/microbiomedata/nmdc-schema/issues/2010) this PR is for @aclum said the 'from' version was 10.3.0

**Comment by eecavanna** [comment]
OK, thanks, @JamesTessmer. I'll work with @aclum and @turbomam tomorrow to identify the version numbers, then I'll put them into this branch.

**Review by turbomam — APPROVED** [review]

**Comment by eecavanna** [comment]
I'll merge this in once the GHA checks pass.
